src/app/PersistentData.h

```diff
@@ -0,0 +1,173 @@
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/core/CHIPPersistentStorageDelegate.h>
+#include <lib/core/CHIPTLV.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+
+namespace chip {
+
+constexpr size_t kPersistentBufferMax = 128;
+
+template <size_t kMaxSerializedSize>
+struct PersistentData
+{
+    virtual ~PersistentData() = default;
+
+    virtual CHIP_ERROR UpdateKey(DefaultStorageKeyAllocator & key) = 0;
+    virtual CHIP_ERROR Serialize(TLV::TLVWriter & writer) const    = 0;
+    virtual CHIP_ERROR Deserialize(TLV::TLVReader & reader)        = 0;
+    virtual void Clear()                                           = 0;
+
+    virtual CHIP_ERROR Save(PersistentStorageDelegate * storage)
+    {
+        VerifyOrReturnError(nullptr != storage, CHIP_ERROR_INVALID_ARGUMENT);
+
+        uint8_t buffer[kMaxSerializedSize] = { 0 };
+        DefaultStorageKeyAllocator key;
+        // Update storage key
+        ReturnErrorOnFailure(UpdateKey(key));
+
+        // Serialize the data
+        TLV::TLVWriter writer;
+        writer.Init(buffer, sizeof(buffer));
+        ReturnErrorOnFailure(Serialize(writer));
+
+        // Save serialized data
+        return storage->SyncSetKeyValue(key.KeyName(), buffer, static_cast<uint16_t>(writer.GetLengthWritten()));
+    }
+
+    CHIP_ERROR Load(PersistentStorageDelegate * storage)
+    {
+        VerifyOrReturnError(nullptr != storage, CHIP_ERROR_INVALID_ARGUMENT);
+
+        uint8_t buffer[kMaxSerializedSize] = { 0 };
+        DefaultStorageKeyAllocator key;
+
+        // Set data to defaults
+        Clear();
+
+        // Update storage key
+        ReturnErrorOnFailure(UpdateKey(key));
+
+        // Load the serialized data
+        uint16_t size  = static_cast<uint16_t>(sizeof(buffer));
+        CHIP_ERROR err = storage->SyncGetKeyValue(key.KeyName(), buffer, size);
+        VerifyOrReturnError(CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND != err, CHIP_ERROR_NOT_FOUND);
+        ReturnErrorOnFailure(err);
+
+        // Decode serialized data
+        TLV::TLVReader reader;
+        reader.Init(buffer, size);
+        return Deserialize(reader);
+    }
+
+    virtual CHIP_ERROR Delete(PersistentStorageDelegate * storage)
+    {
+        VerifyOrReturnError(nullptr != storage, CHIP_ERROR_INVALID_ARGUMENT);
+
+        DefaultStorageKeyAllocator key;
+        // Update storage key
+        ReturnErrorOnFailure(UpdateKey(key));
+        // Delete stored data
+        return storage->SyncDeleteKeyValue(key.KeyName());
+    }
+};
+
+struct FabricList : public PersistentData<kPersistentBufferMax>
```

Comment:
Why is FabricList in PeristentData? Should we split each type into a separate header?